doc/iterations/2025.1/add_support_for_model_publication.adoc

```diff
@@ -0,0 +1,80 @@
+= Add support for model publication
+
+== Problem
+
+Sirius Web doesn't provide an **in-application** way to import models from other projects.
+The current import/export mechanisms rely on generating an interchange format file (Json, XMI, SysML), and uploading such file in a project to import it.
+
+When exported models are used as shared reusable components, it is the responsibility of the stakeholders to:
+
+- Store the shared models in an external location
+- Ensure that every user is working on the same version of the shared model
+- Coordinate to export a new version when needed, and import it in all the dependent projects
+
+These processes are error-prone, and could be more controlled if they all happened inside the application.
+
+== Key Result
+
+Sirius Web users should be able to:
+
+- Publish a model with a given version number
+- Import a published with a given version number
+
+== Solution
+
+The contextual menu of _Model_ elements in the explorer contains a new _Publish_ tool.
```

Comment:
If we want to publish parts of a project or even better if we want to publish multiple artifacts for a project, then I'm not sure that Sirius Web can decide how to do it. Think about how we are exporting X maven artifacts (that may be preprocessed by the way) from one repository. It then creates a dependency tree that is used by the one using the dependency.

Reply:
After having thought about this a bit more, given the fact that we have multiple downstream projects (at least 3 I think) that have implemented something along those lines, I suspect that we can't decide on our own what publishing means (for example a validation step before the publication, some processing to transform the data, etc).

I suspect that we should be able to register "workflows" (think Github Workflows for Github Actions) which let specifiers define how to publish things for a project.

```
public interface IProjectWorkflow {
    /*
     * Can us the nature of the project to figure out if the workflow is relevant
     */
    boolean canHandle(UUID projectId);

     /*
      * I'm not providing the editing context here on purpose to let the specifier load the snapshot of the
      * data from the database in order NOT to block the editing context event processor.
      */
    void handle(UUID projectId, ...);
}
```

A project could have multiple workflows (like we do in our Github repository) and the end user could just select the workflow to run for the publication. We have to keep in mind that a workflow could take some time (a few milliseconds to a couple of hours). The workflow would explicitly decide to perform a publication (it may not publish anything at all). One could make a workflow to publish everything, or another one which would for example only publish the `papaya::Component` from a Papaya project and not other artifacts.

If you want to run a workflow using some variables as input (here a version, or a specific object) you may contribute a dedicated tool to execute the mutation triggering the workflow in your own application. Some downstream projects already have dedicated user interface accessible conditionally from specific objects in specific projects to start this. In such use case, all downstream projects would "just" share a common workflow execution system and publication area (think Github Packages).

Reply:
So there are probably two shapes for this topic:

- How to publish artifacts
- How to consume artifacts (project dependencies, reference vs copy etc)

Reply:
I see, if Sirius Web forces specifiers to publish entire editing contexts, which is, as you said, the only thing that makes sense in a vacuum, this may conflict with existing publication implementations. For example, I know some downstream application have a publication mechanism where only a part of a model is published, because the application ensures that this sub model does not depend on anything.

If we go for the workflow approach, should Sirius Web provide a specific table for that? I am not sure this is what you mean by publication area. If we have such table in Sirius Web we need to let downstream applications add information next to the published model (e.g. version number, date, etc). This could be their responsibility, but in this case they would have to do kind of a lot of work every time they want to re-use the publication mechanism of Sirius Web. Maybe we can provide some default publication data that can be easily reused?

I agree this could be split into two shapes.

Reply:
> If we go for the workflow approach, should Sirius Web provide a specific table for that? I am not sure this is what you mean by publication area.

I think that there be should somewhere (probably another page) where users could see what has been published by a project. It would just list the artifacts published and when they have been published. Using a simple "static" table in a dedicated page (something like `/projects/:projectId/artifacts` or something.

> If we have such table in Sirius Web we need to let downstream applications add information next to the published model (e.g. version number, date, etc). This could be their responsibility, but in this case they would have to do kind of a lot of work every time they want to re-use the publication mechanism of Sirius Web. Maybe we can provide some default publication data that can be easily reused?

I was thinking about that and for example in Papyrus Web (since I worked on this one a tiny bit), they have the following properties:

- version kind (development, release, major or custom)
- version (can be computed from version kind)
- date
- author
- comments
- copyright
- message

Some fields are computed from others or from the backend for example they can check the latest artifact published to figure out the previous version and offer to create a new upgraded version (using a new release version kind to automatically go from 1.2.3 to 1.3.0 for example).

I don't think that every downstream projects could use the same properties and even if all artifacts on an instance of Sirius Web could use the same properties. They would still be free to provide a custom user interface to publish models but we could provide a way for specifiers to add some properties and tell us how to display them.

Something like this:

```
package org.eclipse.sirius.web.domain.boundedcontexts.publication;

@Table("artifact")
public class Artifact extends AbstractValidatingAggregateRoot<Artifact> implements Persistable<UUID> {
    @Transient
    private boolean isNew;

    @Id
    private UUID id;

    private String version;

    private String description;

    @MappedCollection(idColumn = "artifact_id", keyColumn = "index")
    private List<ArtifactProperty> properties;

    private Instant createdOn;

    private Instant lastModifiedOn;
}
```

with something like this:

```
@Table("artifact_property")
public record ArtifactProperty(
    String name,
    String value
) {}
```

This way, if a downstream application wants to provide a custom user interface or just a custom workflow which would capture additional properties to be attached to the artifact published, they could tell us. On the frontend, we would only, for now display the properties in additional columns of the table. The field version and description would be mandatory (the description could be blank but at least it would be handled natively by Sirius Web) and additional fields could be contributed if a downstream project wants it. This part regarding custom properties of the artifact published would definitely be a cutting back that could be implemented later.